### PR TITLE
Label pods with the revision of the control plane that injected them

### PR DESCRIPTION
**Pull request by esnible**
As part of the work on https://github.com/istio/istio/issues/22274 we label injected pods with the control plane revision that injected them.

In the future the `istioctl` troubleshooting commands will use this to determine which revision of the control plane is associated with a pod.

### CODE REVIEW

File: pkg/kube/inject/inject.go
Comment (howardjohn) [inline pkg/kube/inject/inject.go]: Does this need to be a label as opposed to an annotation?

Comment (esnible) [reply pkg/kube/inject/inject.go]: Your choice.

I thought it would be cool to be able to say `kubectl get pods --all-namespaces --selector istio.io/rev=canary`.  Where the revision hangs in the metadata should not affect my planned control plane finding implementation for istioctl.

**Comment by howardjohn** [comment]
Actually I agree, label is better. That exact use case was called out in
the original doc. Thanks!

File: manifests/istio-control/istio-discovery/files/gen-istio.yaml
Comment (liamawhite) [inline manifests/istio-control/istio-discovery/files/gen-istio.yaml]: Is this field named revision as the idea would be users would have two different versions of istiod running?

Comment (esnible) [reply manifests/istio-control/istio-discovery/files/gen-istio.yaml]: Currently the charts leave the `revision` blank by default and for the case we have only one control plane.

I feel this is confusing; I would prefer "revision" to default to `global` or to work for example "docker" which assigns a default name such as `hyperactive-marx`.

We could suppress `revision` label injection if there is no revision.  I did it this way because I think it is simpler, and also because I was thinking of the case where the user installs 1.6.0 without a revision but then installs 1.6.1 with revision `canary`.  It would be confusing to see the label on some pods but not others.

**Comment by esnible** [comment]
@ostromart Are you familiar with the template linting?  I see an error 
```
==> Linting manifests/istio-control/istio-discovery
[ERROR] templates/: render error in "istio-discovery/templates/istiod-injector-configmap.yaml": template: istio-discovery/templates/istiod-injector-configmap.yaml:9:21: executing "istio-discovery/templates/istiod-injector-configmap.yaml" at <.Revision>: map has no entry for key "Revision"
```
and don't know which map needs the key to pass the linting.

File: manifests/istio-control/istio-discovery/files/gen-istio.yaml
Comment (howardjohn) [inline manifests/istio-control/istio-discovery/files/gen-istio.yaml]: I think we don't need a new field here, the sidecar injector already knows what revision this is

Comment (esnible) [reply manifests/istio-control/istio-discovery/files/gen-istio.yaml]: Where?  On my install from master the injector is istiod.  The istiod-global pod has a label `version: global` and I have requested that @istio/wg-environments-maintainers add `istio.io/rev: global`.  But there is no `env.name` = `valueFrom.fieldRef` to get it into istiod.

It might be better to tell istiod its Revision instead of putting it here.  Shall I modify this PR to do that?

Comment (esnible) [reply manifests/istio-control/istio-discovery/files/gen-istio.yaml]: Fixed.

**Comment by esnible** [comment]
@liamawhite @howardjohn Please re-review.

@howardjohn For istioctl, this PR uses the "Env" Revision, so the user currently needs to do `REVISION=canary istioctl kube-inject -f <file>`.  Please approve this anyway; if there is a need for `--revision` on `kube-inject` we will add it as a follow-on.  The most important case is for injection.

File: manifests/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml
Comment (howardjohn) [inline manifests/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml]: This should exist on master and I think will conflict as I have it as `{{ .Values.revision | default "default" }}`. Probably needs a rebase?

File: pkg/kube/inject/inject.go
Comment (howardjohn) [inline pkg/kube/inject/inject.go]: We should not use `os.Getenv()`. There is supposed to be a linter for this so if its not failing that must be broken. Can you reference `revisionVar` in pilot/pkg/bootstrap (and make it `RevisionVar`)?

File: pkg/kube/inject/webhook.go
Comment (howardjohn) [inline pkg/kube/inject/webhook.go]: same comment about env

Comment (howardjohn) [reply pkg/kube/inject/webhook.go]: We should probably default to `default` like we do in the install charts

**Review by howardjohn — COMMENTED** [review]
Thanks for working on this! LGTM in general, just a few comments

**Review by liamawhite — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
